File: LeetCode/114FlattenBinaryTreeToLinkedList/FlattenBinaryTreeToLinkedList.py
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        # DFS -> Time:O(n), Space:O(h)
        def dfs(root):
            if not root:
                return None
            leftTail = dfs(root.left)
            rightTail = dfs(root.right)
            if leftTail:
                leftTail.right = root.right
                root.right = root.left
                root.left = None
            last = rightTail or leftTail or root
            return last
        dfs(root)

File: LeetCode/114FlattenBinaryTreeToLinkedList/FlattenBinaryTreeToLinkedList.cpp
class Solution {
public:
    TreeNode* dfs(TreeNode* root) {
        if (root == NULL) {
            return NULL;
        }
        TreeNode* leftTail = dfs(root->left);
        TreeNode* rightTail = dfs(root->right);
        if (leftTail) {
            leftTail->right = root->right;
            root->right = root->left;
            root->left = NULL;
        }
        if (rightTail) {
            return rightTail;
        }
        else if (leftTail) {
            return leftTail;
        }
        return root;
    }
    void flatten(TreeNode* root) {
        // DFS -> Time:O(n), Space:O(h)
        dfs(root);
    }
};
